ibis/expr/types/relations.py

```diff
@@ -917,40 +917,173 @@ def aggregate(
+        ├───────────┼───────────┼────────────────┼──────────────┼───────────────────┼──┤
+        │ Adelie    │ Torgersen │           43.1 │         19.2 │               197 │  │
+        │ Adelie    │ Biscoe    │           42.7 │         18.3 │               196 │  │
+        │ Adelie    │ Dream     │           41.5 │         18.5 │               201 │  │
+        │ Gentoo    │ Biscoe    │           49.9 │         16.1 │               213 │  │
+        │ Chinstrap │ Dream     │           50.2 │         18.7 │               198 │  │
+        └───────────┴───────────┴────────────────┴──────────────┴───────────────────┴──┘
+
+        Drop all duplicated rows
+
+        >>> expr = t.distinct(on=["species", "island", "year", "bill_length_mm"], keep=None)
+        >>> expr.count()
+        273
+        >>> t.count()
+        344
+
+        You can pass [`selectors`][ibis.expr.selectors] to `on`
+
+        >>> t.distinct(on=~s.numeric())
+        ┏━━━━━━━━━┳━━━━━━━━━━━┳━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┳━━━┓
+        ┃ species ┃ island    ┃ bill_length_mm ┃ bill_depth_mm ┃ flipper_length_mm ┃ … ┃
+        ┡━━━━━━━━━╇━━━━━━━━━━━╇━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━╇━━━┩
+        │ string  │ string    │ float64        │ float64       │ int64             │ … │
+        ├─────────┼───────────┼────────────────┼───────────────┼───────────────────┼───┤
+        │ Adelie  │ Torgersen │           39.1 │          18.7 │               181 │ … │
+        │ Adelie  │ Torgersen │           39.5 │          17.4 │               186 │ … │
+        │ Adelie  │ Torgersen │            nan │           nan │                 ∅ │ … │
+        │ Adelie  │ Biscoe    │           37.8 │          18.3 │               174 │ … │
+        │ Adelie  │ Biscoe    │           37.7 │          18.7 │               180 │ … │
+        │ Adelie  │ Dream     │           39.5 │          16.7 │               178 │ … │
+        │ Adelie  │ Dream     │           37.2 │          18.1 │               178 │ … │
+        │ Adelie  │ Dream     │           37.5 │          18.9 │               179 │ … │
+        │ Gentoo  │ Biscoe    │           46.1 │          13.2 │               211 │ … │
+        │ Gentoo  │ Biscoe    │           50.0 │          16.3 │               230 │ … │
+        │ …       │ …         │              … │             … │                 … │ … │
+        └─────────┴───────────┴────────────────┴───────────────┴───────────────────┴───┘
+
+        The only valid values of `keep` are `"first"`, `"last"` and [`None][None]
+
+        >>> t.distinct(on="species", keep="second")
+        Traceback (most recent call last):
+          ...
+        ibis.common.exceptions.IbisError: Invalid value for keep: 'second' ...
         """
-        return ops.Distinct(self).to_expr()
+
+        import ibis.expr.selectors as s
+
+        if on is None:
+            # dedup everything
+            if keep != "first":
```

Comment:
wouldn't keep="last" also makes sense? It would just give you the same result as "first", but it still makes sense.

keep=None also would make sense, but to support that we would need to update the Distinct op.

Reply:
`keep=None` can be supported by dropping the `DISTINCT` operator entirely and just using group by agg.

A `GROUP BY` with no aggregates is a valid operation that is effectively a `DISTINCT` so that might be the way to go here.

Reply:
I probably won't implement that in this PR though.